Add `hy_goal_favor` and `hy_goal_disfavor`

jlebon opened the pull request:
Minimal wrappers around `SOLVER_FAVOR` and `SOLVER_DISFAVOR`. We'd like
to make use of these in rpm-ostree.

jlebon committed:
Add `hy_goal_favor` and `hy_goal_disfavor`

Minimal wrappers around `SOLVER_FAVOR` and `SOLVER_DISFAVOR`. We'd like
to make use of these in rpm-ostree.

jlebon committed:
Define a cleanup function for `DnfPackageSet`

This is the same thing we already have for e.g. `DnfReldepList`. These
helpers are really useful to keep C client code clean and make it harder
to leak resources.

jlebon commented:
Any thoughts on this?

j-mracek commented:
@jlebon Please could you provide information for what purpose (user-case) you want to use solver_favor?

j-mracek commented:
@jlebon Based also on user-case we would like to have some test, to guarantee that it does what is expected.

Conan-Kudo commented:
`SOLVER_FAVOR` (and its opposing counterpart `SOLVER_DISFAVOR`) were added by @mlschroe to libsolv to implement the capability to support urpmi's `prefer.vendor.list` feature in DNF and other package managers. The idea is that you could maintain an externalized list of preferences that the solver would account for that could be dynamically managed without requiring building an empty package with a pile of Suggests to influence the transaction.

jlebon commented:
OK, now added tests for this and added `hy_goal_disfavor` for completeness!

> @jlebon Please could you provide information for what purpose (user-case) you want to use solver_favor?

Yes, this will allow us to have more influence over libsolv's decisions. Without this, it's impossible to make libsolv "prefer" a package but still letting it not be part of a possible solution. For more details on the rpm-ostree use case, see https://github.com/coreos/rpm-ostree/pull/2059. The goal there is for "locked" packages to be given preference when updating the lockfile, but still allow solutions which require picking packages outside the lockfile.

j-mracek commented:
LGTM!

j-mracek commented:
@rh-atomic-bot r+

jlebon commented:
Hmm, hard to tell what failed. I see:

```
1 feature passed, 0 failed, 0 skipped
42 scenarios passed, 0 failed, 0 skipped
168 steps passed, 0 failed, 0 skipped, 0 undefined
Took 0m9.261s
+ :
+ popd
/var/lib/jenkins/workspace/dnf-libdnf-test
[dnf-libdnf-test] $ python /tmp/jenkins2264749669589379953.py
Recording test results
ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error?
```

The last few runs of `dnf-libdnf-test` failed the same way, so I think it's unrelated to this PR?

Ahhh, I *think* it might be related to this:

```
Exception PermissionError: [Errno 13] Permission denied: '/junit/features'
Traceback (most recent call last):
  File "/usr/bin/behave", line 11, in <module>
    load_entry_point('behave==1.2.6', 'console_scripts', 'behave')()
  File "/usr/lib/python3.7/site-packages/behave/__main__.py", line 183, in main
    return run_behave(config)
  File "/usr/lib/python3.7/site-packages/behave/__main__.py", line 127, in run_behave
    failed = runner.run()
  File "/usr/lib/python3.7/site-packages/behave/runner.py", line 804, in run
    return self.run_with_paths()
  File "/usr/lib/python3.7/site-packages/behave/runner.py", line 824, in run_with_paths
    return self.run_model()
  File "/usr/lib/python3.7/site-packages/behave/runner.py", line 640, in run_model
    reporter.feature(feature)
  File "/usr/lib/python3.7/site-packages/behave/reporter/junit.py", line 264, in feature
    os.makedirs(self.config.junit_directory)
  File "/usr/lib64/python3.7/os.py", line 211, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/usr/lib64/python3.7/os.py", line 211, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/usr/lib64/python3.7/os.py", line 221, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/junit/features'
```

j-mracek commented:
Lets try it again

j-mracek commented:
@rh-atomic-bot retry

m-blaha commented:
Sorry guys - misconfiguration in jenkins job. Should by fixed now.

m-blaha commented:
@rh-atomic-bot retry
